src/customfilesystemmodel.cpp

```diff
@@ -82,27 +110,48 @@ bool CustomFileSystemModel::setData(const QModelIndex &index,
             _checklist.insert(index);
             emit dataChanged(index, index, selectionChangedRole);
 
-            QModelIndex parent = index.parent();
-            if(parent.isValid())
+            // Recursively set PartiallyChecked on all ancestors, and make an
+            // ordered list of checked ancestors.
+            QModelIndex        parent         = index.parent();
+            QModelIndex        previousParent = index;
+            QList<QModelIndex> checkedAncestors;
+            while(parent.isValid())
             {
-                if(parent.data(Qt::CheckStateRole) == Qt::Checked)
+                if(hasCheckedAncestor(parent))
```

Comment:
If `parent` is checked, or if any ancestor of `parent` is checked, it returns true.

I'm trying to think of a better name... `isCheckedOrAncestorChecked` is more clear, but it's too long.  Another option would be to write

    if((dataInternal(parent) == Qt::Checked) || (hasCheckedAncestor(parent))

and change `hasCheckedAncestor` to only check ancestors of the given index, but that also seems too verbose.